tortoise/verifying_tortoise.go

```diff
@@ -1208,11 +1203,9 @@ func (t *turtle) heal(ctx context.Context, targetLayerID types.LayerID) {
 		// Note: we look at ALL blocks we've seen for the layer, not just those we've previously marked contextually valid
 		logger.Info("self healing attempting to verify candidate layer")
 
-		layerBlockIds, err := t.bdp.LayerBlockIds(candidateLayerID)
+		layerBlockIds, err := t.getLayerBlocksIDs(ctx, candidateLayerID)
```

Comment:
i dont understand your concern. context is not reused between requests. if blocks were in when self healing started they will be found.

Reply:
it may matter if contextually valid blocks will change, so i will re-init context in that case

Reply:
in self healing, we dont just look at contextual valid blocks, we looked at all blocks, including blocks that arrived late.

so it is possible that we have asked the blocked IDs for this layer and cached it before entering self healing. then a late block arrived after the caching took place. especially during the network rejoin, we want to get as many blocks as possible during self healing. caching here is counter productive

Reply:
ah thats not exactly the case, for late block to take effect it needs to be processed, and every layer after should be re-processed. otherwise, its votes are mostly useless. so what you are describing is only relevant during rerun.

self-healing during normal operation is only about dealing with stuck verified layer due to the absence of good blocks.

Reply:
>  then a late block arrived after the caching took place. especially during the network rejoin, we want to get as many blocks as possible during self healing. caching here is counter productive

also regardless of how it works now, relying on such non-determinism is a mistake anyway

Reply:
> self-healing during normal operation is only about dealing with stuck verified layer due to the absence of good blocks.

i disagree. in the general terms, self-healing is about being able to heal from temporary malicious majority or network partition. which means we want to count votes that are late or use different beacons.

to that end, we should not rely on the in-memory opinion maps to count votes. i agree the current code is as you described, but i disagree we should keep it that way. we should reload/reprocess all the votes from disk.

in any case, i don't see any harm using just plain context here in the healing case. but then i'm ok if you leave it.

Reply:
> we should reload/reprocess all the votes from disk.

we do that during full rerun periodically. in the future, i assume, we will have an interface for triggering rerun based on sync or other conditions. we can do it even now by counting number of late blocks in block handler and trigger rerun, but is it worth it?

i am not really sure that this is viable to reprocess blocks in self-healing unconditionally.

Reply:
> i am not really sure that this is viable to reprocess blocks in self-healing unconditionally.

my understanding is it's by design. but with rerun i am not sure anymore the responsibility of each routine....